### PR TITLE
feat(dashboard-layout): Add layout field to endpoint

### PR DESCRIPTION
**Pull request by narsaynorath**
Exposes `layout` as a field in the API for `/api/0/organizations/org-id/dashboards/` and `/api/0/organizations/sentry/dashboards/id/`.

# Behaviours
## GET /api/0/organizations/sentry/dashboards/id/
* If there is a layout saved in `detail` field, return that, else return None (i.e. no layout, or no detail set)

## POST /api/0/organizations/org-id/dashboards/
* If widgets have layout defined in request, then validate and set them in the `detail` field
* Allows `layout: null` (not default behaviour provided by `JSONField`) 
* Defaults to `None` if not provided at all
* Expects all valid keys to be ints

## PUT /api/0/organizations/sentry/dashboards/id/
* Has the same validation as POST
* If not provided `layout`, does not overwrite the previous value

### CODE REVIEW

File: src/sentry/api/serializers/rest_framework/dashboard.py [hunk 1]
```diff
@@ -152,6 +152,7 @@ class DashboardWidgetSerializer(CamelSnakeSerializer):
     widget_type = serializers.ChoiceField(
         choices=DashboardWidgetTypes.as_text_choices(), required=False
     )
+    layout = serializers.JSONField(required=False, allow_null=True)
```
Comment (shruthilayaj) [inline src/sentry/api/serializers/rest_framework/dashboard.py:155]: I was wondering if we could do some more validation on the incoming JSON data but see that the [JSONField](https://www.django-rest-framework.org/api-guide/fields/#jsonfield) does not accept child serializers, so I think what you have should be fine or we could use DictField which does allow that, but I'm not sure if we need to be restrictive like that (just thinking ahead to all the public api conversations)

File: src/sentry/api/serializers/rest_framework/dashboard.py [hunk 2]
```diff
@@ -166,6 +167,38 @@ def validate_interval(self, interval):
...
+        if invalid_keys:
+            invalid_keys_str = ", ".join(invalid_keys)
+            raise serializers.ValidationError(f"Contains invalid keys: {invalid_keys_str}")
+
+        layout_to_store = {}
+        for key in layout.keys():
+            if key in STORE_KEYS:
+                if not isinstance(layout[key], int):
+                    raise serializers.ValidationError(f"Expected number for: {key}")
+                layout_to_store[key] = layout[key]
```
Comment (shruthilayaj) [inline src/sentry/api/serializers/rest_framework/dashboard.py:199]: nit: can we loop through the STORE_KEYS instead? Esp if we don't end up doing validating the layout field JSON/Dict we may not know what/how many keys are coming through in this layout dict, so probably better to loop through known keys.

```suggestion
        for key in STORE_KEYS:
            if layout.get(key) is None:
                continue

            if not isinstance(layout[key], int):
                raise serializers.ValidationError(f"Expected number for: {key}")
            layout_to_store[key] = layout[key]
```

File: src/sentry/api/serializers/rest_framework/dashboard.py [hunk 3]
```diff
@@ -289,6 +322,7 @@ def create_widget(self, dashboard, widget_data, order):
             if "widget_type" in widget_data
             else DashboardWidgetTypes.DISCOVER,
             order=order,
+            detail={"layout": widget_data.get("layout", None)},
```
Comment (shruthilayaj) [inline src/sentry/api/serializers/rest_framework/dashboard.py:325]: Defaults to None
```suggestion
            detail={"layout": widget_data.get("layout")},
```

File: src/sentry/api/serializers/models/dashboard.py [hunk 4]
```diff
@@ -42,6 +42,7 @@ def serialize(self, obj, attrs, user, **kwargs):
             # Default to discover type if null
             "widgetType": DashboardWidgetTypes.get_type_name(obj.widget_type)
             or DashboardWidgetTypes.TYPE_NAMES[0],
+            "layout": obj.detail["layout"] if obj.detail else None,
```
[the remark this answers is not in the document]
Comment (narsaynorath) [reply src/sentry/api/serializers/models/dashboard.py]: https://github.com/getsentry/sentry/pull/31050/files#diff-c1b7e86a0bab988138a71df37b4123f3714c98be1bc4194bc7501f815ba66d32R45

File: src/sentry/api/serializers/rest_framework/dashboard.py (hunk 1 above)
Comment (narsaynorath) [reply src/sentry/api/serializers/rest_framework/dashboard.py]: https://github.com/getsentry/sentry/pull/31050/files#diff-016a07238f9714dd8918a63f4930f5dec5109e0a20b1807ad5ed98075666e6bcR52

I decided to split it off into a custom field where I moved the validation. Let me know if you agree with this change, or we can look into the DictField!

With my current understanding, using a DictField would involve adding the field, specifying a child serializer (IntegerField?), and doing the key validation in the DashboardWidgetSerializer (i.e. kind of the way it is now). I'm concerned about supporting different value types (e.g. there are some boolean fields with RGL, we just don't have a use for them yet) and my thoughts on separating the validation logic to a custom field means there's less code in the `DashboardWidgetSerializer` this isn't directly related to validating a dashboard widget.

File: src/sentry/api/serializers/rest_framework/dashboard.py (hunk 2 above)
Comment (narsaynorath) [reply src/sentry/api/serializers/rest_framework/dashboard.py]: https://github.com/getsentry/sentry/pull/31050/files#diff-016a07238f9714dd8918a63f4930f5dec5109e0a20b1807ad5ed98075666e6bcR69

Thanks for the suggestion!

File: src/sentry/api/serializers/rest_framework/dashboard.py [hunk 5]
```diff
@@ -305,11 +335,13 @@ def create_widget(self, dashboard, widget_data, order):
         DashboardWidgetQuery.objects.bulk_create(new_queries)
 
     def update_widget(self, widget, data, order):
+        prev_layout = widget.detail.get("layout", None) if widget.detail else None
         widget.title = data.get("title", widget.title)
         widget.display_type = data.get("display_type", widget.display_type)
         widget.interval = data.get("interval", widget.interval)
         widget.widget_type = data.get("widget_type", widget.widget_type)
         widget.order = order
+        widget.detail = {"layout": data.get("layout", prev_layout)}
```
Comment (shruthilayaj) [inline src/sentry/api/serializers/rest_framework/dashboard.py:344]: Updating widget.detail might start to get a bit ugly as we add more keys to the `detail` dict. Wondering if we need access to the widget instance and have to get prev_layout? If we should do something like in the validate() of [DiscoverSavedQuery](https://github.com/getsentry/sentry/blob/03ee96982b9e86bce623f2c518f7ee6b1e9bf6ba/src/sentry/discover/endpoints/serializers.py#L225-L230) where we build the detail dict in the validate? 

What we have might be fine for now though.

File: src/sentry/api/serializers/rest_framework/dashboard.py [hunk 6]
```diff
@@ -305,11 +335,13 @@ def create_widget(self, dashboard, widget_data, order):
         DashboardWidgetQuery.objects.bulk_create(new_queries)
 
     def update_widget(self, widget, data, order):
+        prev_layout = widget.detail.get("layout", None) if widget.detail else None
```
Comment (shruthilayaj) [inline src/sentry/api/serializers/rest_framework/dashboard.py:338]: nit
```suggestion
        prev_layout = widget.detail.get("layout") if widget.detail else None
```

**Review by shruthilayaj — APPROVED** [review]

File: src/sentry/api/serializers/rest_framework/dashboard.py (hunk 5 above)
Comment (narsaynorath) [reply src/sentry/api/serializers/rest_framework/dashboard.py:348]: I agree that it's going to start looking messy when we start adding more fields.

This line is covered by the `test_update_without_specifying_layout_does_not_change_saved_layout` test. Now I'm wondering if this behaviour is too flexible. If we don't grab `prev_layout` then we could just default the layout to `None`.

At the time I was thinking of the behaviour not just from a UI usage, but a general use case and avoiding unintended destructive behaviour, seeing as how the rest of this function does something similar.

Comment (narsaynorath) [reply src/sentry/api/serializers/rest_framework/dashboard.py:348]: We chatted about this on a call. In summary we decided to keep this as-is and conduct a separate discussion on the behaviour of endpoints. Our concerns here are that this is a PUT request for a dashboard, but this almost acts like a PATCH for widgets. It is slightly confusing what defaults we should have in a case of nested serializers like this.

**Pull request merged**